### PR TITLE
Linkage Checker to add ID to repositories

### LINKED ISSUE #1656: LinkageCheckerRule fails to resolve grpc dependencies
https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/49 failed.

The problem occurs when we clear the local Maven repository cache:

```
rm -rf ~/.m2/repository/io
```

and run:

```
./mvnw \
            --activate-profiles linkage-check \
            --batch-mode \
            --show-version \
            --threads 1.5C \
            --define maven.test.skip=true \
            --define skip.surefire.tests=true \
            --define skipITs=true \
            --define maven.javadoc.skip=true \
            --errors \
            verify
```

I get

```
Failed to collect dependency: [
io.grpc:grpc-api:jar:1.26.0 was not resolved. Dependency path: null > com.google.cloud:google-cloud-notification:jar:0.120.2-beta (compile) > com.google.cloud:google-cloud-pubsub:jar:1.102.1 (compile) > io.grpc:grpc-api:jar:1.26.0 (compile),
io.grpc:grpc-api:jar:1.25.0 was not resolved. Dependency path: null > com.google.cloud:google-cloud-notification:jar:0.120.2-beta (compile) > com.google.cloud:google-cloud-pubsub:jar:1.102.1 (compile) > com.google.api:gax-grpc:jar:1.52.0 (compile) > io.grpc:grpc-stub:jar:1.25.0 (compile) > io.grpc:grpc-api:jar:1.25.0 (compile),
io.grpc:grpc-core:jar:1.28.1 was not resolved. Dependency path: null > com.google.cloud:spring-cloud-gcp-vision:jar:2.0.0-SNAPSHOT (compile) > com.google.cloud:google-cloud-vision:jar:1.99.3 (compile) > com.google.api:gax-grpc:jar:1.56.0 (compile) > io.grpc:grpc-netty-shaded:jar:1.28.1 (compile) > io.grpc:grpc-core:jar:1.28.1 (compile),
io.grpc:grpc-api:jar:1.30.2 was not resolved. Dependency path: null > com.google.cloud:spring-cloud-gcp-pubsub:jar:2.0.0-SNAPSHOT (compile) > com.google.cloud:google-cloud-pubsub:jar:1.108.0 (compile) > io.grpc:grpc-api:jar:1.30.2 (compile),
io.grpc:grpc-core:jar:1.30.2 was not resolved. Dependency path: null > com.google.cloud:spring-cloud-gcp-pubsub:jar:2.0.0-SNAPSHOT (compile) > com.google.cloud:google-cloud-pubsub:jar:1.108.0 (compile) > io.grpc:grpc-core:jar:1.30.2 (compile),
io.grpc:grpc-api:jar:1.28.1 was not resolved. Dependency path: null > com.google.cloud:spring-cloud-gcp-vision:jar:2.0.0-SNAPSHOT (compile) > com.google.cloud:google-cloud-vision:jar:1.99.3 (compile) > io.grpc:grpc-api:jar:1.28.1 (compile),
io.grpc:grpc-netty-shaded:jar:1.30.2 was not resolved. Dependency path: null > com.google.cloud:spring-cloud-gcp-pubsub:jar:2.0.0-SNAPSHOT (compile) > com.google.cloud:google-cloud-pubsub:jar:1.108.0 (compile) > io.grpc:grpc-netty-shaded:jar:1.30.2 (compile),
io.grpc:grpc-api:jar:1.27.2 was not resolved. Dependency path: null > com.google.api.grpc:grpc-google-cloud-redis-v1beta1:jar:0.85.0 (compile) > io.grpc:grpc-api:jar:1.27.2 (compile),
io.grpc:grpc-api:jar:1.30.0 was not resolved. Dependency path: null > com.google.cloud:spring-cloud-gcp-data-datastore:jar:2.0.0-SNAPSHOT (compile) > com.google.cloud:google-cloud-datastore:jar:1.104.0 (compile) > io.grpc:grpc-api:jar:1.30.0 (compile),
io.grpc:grpc-netty-shaded:jar:1.28.1 was not resolved. Dependency path: null > com.google.cloud:spring-cloud-gcp-vision:jar:2.0.0-SNAPSHOT (compile) > com.google.cloud:google-cloud-vision:jar:1.99.3 (compile) > com.google.api:gax-grpc:jar:1.56.0 (compile) > io.grpc:grpc-netty-shaded:jar:1.28.1 (compile),
io.grpc:grpc-core:jar:1.25.0 was not resolved. Dependency path: null > com.google.cloud:google-cloud-notification:jar:0.120.2-beta (compile) > com.google.cloud:google-cloud-pubsub:jar:1.102.1 (compile) > com.google.api:gax-grpc:jar:1.52.0 (compile) > io.grpc:grpc-netty-shaded:jar:1.25.0 (compile) > io.grpc:grpc-core:jar:1.25.0 (compile),
io.grpc:grpc-netty-shaded:jar:1.27.2 was not resolved. Dependency path: null > com.google.cloud:google-cloud-redis:jar:1.0.0 (compile) > com.google.api:gax-grpc:jar:1.54.0 (compile) > io.grpc:grpc-netty-shaded:jar:1.27.2 (compile),
io.netty:netty-codec-http2:jar:4.1.48.Final was not resolved. Dependency path: null > io.grpc:grpc-netty:jar:1.30.2 (compile) > io.netty:netty-codec-http2:jar:4.1.48.Final (compile),
io.grpc:grpc-netty-shaded:jar:1.25.0 was not resolved. Dependency path: null > com.google.cloud:google-cloud-notification:jar:0.120.2-beta (compile) > com.google.cloud:google-cloud-pubsub:jar:1.102.1 (compile) > com.google.api:gax-grpc:jar:1.52.0 (compile) > io.grpc:grpc-netty-shaded:jar:1.25.0 (compile),
io.grpc:grpc-api:jar:1.29.0 was not resolved. Dependency path: null > com.google.api.grpc:grpc-google-common-protos:jar:1.18.0 (compile) > io.grpc:grpc-api:jar:1.29.0 (compile),
io.grpc:grpc-netty-shaded:jar:1.27.0 was not resolved. Dependency path: null > com.google.cloud:google-cloud-irm:jar:0.2.1 (compile) > com.google.api:gax-grpc:jar:1.53.1 (compile) > io.grpc:grpc-netty-shaded:jar:1.27.0 (compile),
io.grpc:grpc-api:jar:1.27.0 was not resolved. Dependency path: null > com.google.api.grpc:grpc-google-cloud-irm-v1alpha2:jar:0.2.1 (compile) > io.grpc:grpc-api:jar:1.27.0 (compile),
io.grpc:grpc-core:jar:1.27.0 was not resolved. Dependency path: null > com.google.cloud:google-cloud-irm:jar:0.2.1 (compile) > com.google.api:gax-grpc:jar:1.53.1 (compile) > io.grpc:grpc-netty-shaded:jar:1.27.0 (compile) > io.grpc:grpc-core:jar:1.27.0 (compile),
io.grpc:grpc-core:jar:1.27.2 was not resolved. Dependency path: null > com.google.cloud:google-cloud-redis:jar:1.0.0 (compile) > com.google.api:gax-grpc:jar:1.54.0 (compile) > io.grpc:grpc-netty-shaded:jar:1.27.2 (compile) > io.grpc:grpc-core:jar:1.27.2 (compile),
io.grpc:grpc-netty-shaded:jar:1.30.0 was not resolved. Dependency path: null > com.google.cloud:spring-cloud-gcp-logging:jar:2.0.0-SNAPSHOT (compile) > com.google.cloud:google-cloud-logging-logback:jar:0.118.0-alpha (compile) > io.grpc:grpc-netty-shaded:jar:1.30.0 (compile),
io.grpc:grpc-core:jar:1.30.0 was not resolved. Dependency path: null > com.google.cloud:spring-cloud-gcp-logging:jar:2.0.0-SNAPSHOT (compile) > com.google.cloud:google-cloud-logging-logback:jar:0.118.0-alpha (compile) > io.grpc:grpc-core:jar:1.30.0 (compile)]
```

I'm suspecting it is caused by version conflicts when building a dependency graph. If that's the case, Linkage Checker should report so.

# Detailed Exception at DependencyGraphBuilder.buildDependencyGraph

```
org.eclipse.aether.resolution.DependencyResolutionException: The following artifacts could not be resolved:
io.netty:netty-codec-http2:jar:4.1.48.Final, io.grpc:grpc-core:jar:1.25.0, io.grpc:grpc-core:jar:1.25.0,
io.grpc:grpc-core:jar:1.25.0, io.grpc:grpc-core:jar:1.27.0, io.grpc:grpc-core:jar:1.27.0, io.grpc:grpc-core:jar:1.27.0,
io.grpc:grpc-core:jar:1.27.2, io.grpc:grpc-core:jar:1.27.2, io.grpc:grpc-core:jar:1.27.2:
 Could not find artifact io.netty:netty-codec-http2:jar:4.1.48.Final in  (https://repo.spring.io/snapshot)
```

But LinkageCheckerRule's repositoryUrls holds https://repo.maven.apache.org/maven2:

```
repositoryUrls = {RegularImmutableList@15343}  size = 4
 0 = "https://repo.spring.io/snapshot"
 1 = "https://repo.spring.io/milestone"
 2 = "https://repo.spring.io/release"
 3 = "https://repo.maven.apache.org/maven2"
```

It does have the artifact:

https://repo.maven.apache.org/maven2/io/netty/netty-codec-http2/4.1.48.Final/

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #1656 .

### CODE REVIEW

**Comment by suztomo** [comment]
Why doesn't it fail...

**Comment by suztomo** [comment]
Good. The test failed in Mac and Windows.

```
[ERROR] testConfigureAdditionalMavenRepositories_resolvingMultipleArtifacts(com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilderTest)  Time elapsed: 1.988 s  <<< FAILURE!
com.google.common.truth.AssertionErrorWithFacts: 
expected to be empty
but was: [io.grpc:grpc-api:jar:1.21.0 was not resolved. Dependency path: null > io.grpc:grpc-core:jar:1.21.0 (compile) > io.grpc:grpc-api:jar:1.21.0 (compile)]
	at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilderTest.testConfigureAdditionalMavenRepositories_resolvingMultipleArtifacts(DependencyGraphBuilderTest.java:394)
```

I'll work on the solution.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java [hunk 1]
```diff
@@ -283,7 +283,7 @@ public static RemoteRepository mavenRepositoryFromUrl(String mavenRepositoryUrl)
     }
 
     RemoteRepository repository =
-        new RemoteRepository.Builder(null, "default", mavenRepositoryUrl).build();
+        new RemoteRepository.Builder(mavenRepositoryUrl, "default", mavenRepositoryUrl).build();
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java:286]: RemoteRepository's ID (the first argument) is nullable as per its Javadoc. However the Maven resolver requires different repositories to have different IDs.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java [hunk 2]
```diff
@@ -370,6 +370,30 @@ public void testConfigureAdditionalMavenRepositories_addingGoogleAndroidReposito
     assertNotNull(graph);
   }
 
+  @Test
+  public void testConfigureAdditionalMavenRepositories_resolvingMultipleArtifacts() {
+    // To verify the effect of the test, you need to cleanup your local Maven repository
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java:375]: This is iffy. The test should not depend on the state of the local maven repository.

Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java]: Moving this to DependencyGraphIntegrationTest

**Review by elharo — APPROVED** [review]

**Comment by suztomo** [comment]
Debugging what the "suppressed exceptions" means in the Windows build failure

```
[ERROR] testArtifactResolutionInDifferentRepository(com.google.cloud.tools.opensource.dependencies.DependencyGraphIntegrationTest)  Time elapsed: 0.375 s  <<< ERROR!
java.nio.file.FileSystemException: C:\Users\builder\.m2\repository\io\grpc: failed to delete one or more files; see suppressed exceptions for details
	at com.google.cloud.tools.opensource.dependencies.DependencyGraphIntegrationTest.testArtifactResolutionInDifferentRepository(DependencyGraphIntegrationTest.java:169)
```

**Comment by suztomo** [comment]
```
	Suppressed: java.nio.file.FileSystemException: C:\Users\builder\.m2\repository\io\grpc\grpc-context\1.17.1\grpc-context-1.17.1.jar: The process cannot access the file because it is being used by another process.
```

**Pull request merged**